File: apps/api/internal/domain/purchase_order.go
package domain

import "time"

// Purchase order statuses. Only draft, ordered and received are reachable
// through the API today.
const (
	PurchaseOrderStatusDraft             = "draft"
	PurchaseOrderStatusOrdered           = "ordered"
	PurchaseOrderStatusPartiallyReceived = "partially_received"
	PurchaseOrderStatusReceived          = "received"
	PurchaseOrderStatusCancelled         = "cancelled"
)

// PurchaseOrder is an order of medicines from a supplier.
type PurchaseOrder struct {
	ID         string
	SupplierID string
	Status     string
	// OrderDate and ExpectedDate are calendar dates; only their year, month and
	// day are meaningful.
	OrderDate    time.Time
	ExpectedDate *time.Time
	CreatedBy    string
	Notes        *string
	CreatedAt    time.Time
	UpdatedAt    time.Time
	// Items and Receipts are only filled by the repository's Create and FindByID.
	Items    []*PurchaseOrderItem
	Receipts []*PurchaseOrderReceipt
}

// PurchaseOrderItem is one medicine and quantity on a purchase order. It is
// immutable once created.
type PurchaseOrderItem struct {
	ID              string
	PurchaseOrderID string
	MedicineID      string
	QuantityOrdered int
	CreatedAt       time.Time
}

// PurchaseOrderReceipt is one receiving event on a purchase order.
type PurchaseOrderReceipt struct {
	ID              string
	PurchaseOrderID string
	ReceivedBy      string
	ReceivedAt      time.Time
	Notes           *string
	CreatedAt       time.Time
	Items           []*PurchaseOrderReceiptItem
}

// PurchaseOrderReceiptItem is what was received of one purchase order item in
// one receiving event.
type PurchaseOrderReceiptItem struct {
	ID                     string
	PurchaseOrderReceiptID string
	PurchaseOrderItemID    string
	QuantityReceived       int
	QuantityDamaged        int
	QuantityReturned       int
	Notes                  *string
	CreatedAt              time.Time
}

File: apps/app/src/types/purchase-order.ts
export type PurchaseOrderStatus =
  | "draft"
  | "ordered"
  | "partially_received"
  | "received"
  | "cancelled"

export interface PurchaseOrder {
  id: string
  supplier_id: string
  status: PurchaseOrderStatus
  order_date: string
  expected_date: string | null
  created_by: string
  notes: string | null
  created_at: string
  updated_at: string
}

export interface PurchaseOrderItem {
  id: string
  medicine_id: string
  quantity_ordered: number
  created_at: string
}

export interface PurchaseOrderReceiptItem {
  id: string
  purchase_order_item_id: string
  quantity_received: number
  quantity_damaged: number
  quantity_returned: number
  notes: string | null
  created_at: string
}

export interface PurchaseOrderReceipt {
  id: string
  purchase_order_id: string
  received_by: string
  received_at: string
  notes: string | null
  created_at: string
  items: PurchaseOrderReceiptItem[]
}

export interface PurchaseOrderDetail extends PurchaseOrder {
  items: PurchaseOrderItem[]
  receipts: PurchaseOrderReceipt[]
}

export interface PurchaseOrderListResponse {
  data: PurchaseOrder[]
  total: number
  limit: number
  offset: number
}

export interface CreatePurchaseOrderItemInput {
  medicine_id: string
  quantity_ordered: number
}

export interface CreatePurchaseOrderInput {
  supplier_id: string
  order_date: string
  expected_date: string | null
  notes: string | null
  items: CreatePurchaseOrderItemInput[]
}

export interface ReceivePurchaseOrderInput {
  notes: string | null
}
